Automattic/wp-calypso
# AppPromo:Add a promo for mobile apps and clean up code to add in unit testing

@johnmaeda asked me to drop in a mobile promo to our current app promo that renders in the reader. The idea was to see this as an experiment to see if we could drive more mobile downloads. I was reluctant to drop the changes into the current code because the current code was untested so I cleaned it up and added some unit tests along the way.

What the changes look like:
<img width="923" alt="screen shot 2016-09-05 at 10 42 42 am" src="https://cloud.githubusercontent.com/assets/437043/18254714/83a818ce-7355-11e6-9539-d48b87e987fd.png">

Notable Changes:
* Extracted promo code logic into it's own component (`./lib/promo-retriever`).
* Extracted dependency on `navigator.userAgent` into it's own component(`/lib/user-agent-utils`). As a note I'll probably follow up with another PR expanding that module and pulling out direct dependencies on navigator.userAgent throughout the codebase.
* I pulled out the direct dependency on analytics and instead attempted to follow the practices outlined [here](https://wpcalypso.wordpress.com/devdocs/client/state/analytics/README.md).
* I extracted showing/hiding logic for the `AppPromo` class and put it into the `ReaderSidebar` which is the only place `AppPromo` renders. Note that I had to tweak the `ReaderSidebar` class a little bit to get it somewhat testable.

To test:
* Unit tests can be run via `npm run test-client client/components/app-promo` and `npm run test-client client/lib/user-agent-utils`.
* To test the component's rendering you'll need to modify `client/reader/sidebar/index.js` to have the `renderAppPromo` method skip all the checks and just return the `AppPromo` component. The promo generation text is random so you'll need to refresh a few times to see the mobile promo text and verify that the hyperlink points to `http://apps.wordpress.com/mobile...`. You'll also want to verify that the desktop links you see point to `http://apps.wordpress.com/desktop...`. If you click on the links you'll need to clear out the localStorage item `desktop_promo_dismissed` in order to see `AppPromo` render again. You can also cheat and just add `show_promo = true` towards the end of the components `getInitialState()`.

**Update** - after chatting with @dmsnell for a little we thought it best to pull the logic for hiding/showing `AppPromo` out of that module and push it into the `ReaderSidebar` class.

p.s. - I haven't written Javascript in six years, anything that looks silly or head bang worthy is probably due to this fact ;).

cc @dmsnell 

Test live: https://calypso.live/?branch=clean-up-app-promo
